[BD-03] [TNL-7349] [BB-2734] Studio UI to edit discussions toggle

Agrendalath opened the pull request:
This adds Studio UI to edit discussions toggle.

It contains changes from #24380, because GitHub doesn't allow changing repository for existing PRs. It will be rebased on `master` after merging the mentioned PR. **Only the last commit should be reviewed for now.**

**JIRA tickets**: TNL-7349

**Dependencies**: #24380

**Screenshots**:
![image](https://user-images.githubusercontent.com/3189670/88421502-f8e3b200-cde8-11ea-91ef-0db72793a25e.png)
![image](https://user-images.githubusercontent.com/3189670/88421529-013bed00-cde9-11ea-98c7-f136a88f4666.png)
![image](https://user-images.githubusercontent.com/3189670/88421565-1153cc80-cde9-11ea-9cad-2f24c232ff81.png)

**Merge deadline**: None, after #24380

**Testing instructions**:

1. Open the course via Studio.
1. Click on the "setttings" button of a unit, enable discussions there.
1. Refresh the page and see that the settings are being retained.
1. Click on the parent subsection's/section's settings and see that the indeterminate status is visible for the discussions there. This is because the discussions are enabled only on some of its children.
1. Check that the recursive enabling/disabling as well as doing nothing and saving works properly.

**Reviewers**
- [ ] @xitij2000
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @Agrendalath! I've created [BLENDED-513](https://openedx.atlassian.net/browse/BLENDED-513) to keep track of it in Jira.

When this pull request is ready, tag your edX technical lead.

<!-- comment:welcome-blended -->
<!-- jenkins ok to test -->

xitij2000 commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -628,6 +625,10 @@ def _save_xblock(user, xblock, data=None, children_strings=None, metadata=None,
             # IMPORTANT NOTE: if the client passed 'null' (None) for a piece of metadata that means 'remove it'. If
             # the intent is to make it None, use the nullout field
             if metadata is not None:
+                discussion_enabled = metadata.pop('discussion_enabled', None)
+                if discussion_enabled is not None:
+                    set_discussion_toggle(discussion_enabled, xblock, user)
+
```
I think these changes should probably be moved to the other PR. Or we can perhaps merge them both into a singe multi-author PR.

Agrendalath replied:
+1 to this, I'd love to see this in #24380. I can extract this to a separate commit for cherry-picking, if needed.

stvstnfrd replied:
I can start looking at the new pieces in this PR, but maybe we can rebase this over the other PR and then update the target/base branch of this PR to be a PR _into_ the other PR branch?

Agrendalath replied:
@stvstnfrd, the problem with this approach is that both PRs are opened from our fork, so AFAIK it's not possible to change the target to this one - we would need to have two different PRs (for our fork and your upstream repository; side effect would be having the comments on only one PR and not having your CI pipelines running). We could also push `open-craft:ahmed/bb-2555-enable-discussion-toggle` to edx remote, but then we would need to keep both branches in sync.

xitij2000 commented on cms/templates/js/discussion-editor.underscore:
```diff
@@ -0,0 +1,26 @@
+        <% if (xblockInfo.isVertical()) { %>
+            <%- gettext('Enable discussion') %>
+        <% } else { %>
+            <%- gettext('Enable discussion for all children') %>
```
```suggestion
            <%- gettext('Enable discussion for all child units') %>
```

xitij2000 commented on cms/templates/js/discussion-editor.underscore:
```diff
@@ -0,0 +1,26 @@
+    <% if (!xblockInfo.isVertical()) { %>
+        <% if (hasDiscussionEnabled === 'partially_enabled') { %>
+            <p class="tip tip-warning">
+                <%- gettext('Currently the discussions are enabled only for some children. If you change the state of the checkbox above, they will be enabled or disabled for all children.') %>
```
```suggestion
                <%- gettext('Currently discussions are enabled only for some children. If you change the state of the checkbox above, they will be enabled or disabled for all child units.') %>
```

stvstnfrd commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -16,7 +16,7 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
         AbstractEditor, BaseDateEditor,
         ReleaseDateEditor, DueDateEditor, GradingEditor, PublishEditor, AbstractVisibilityEditor,
-        StaffLockEditor, UnitAccessEditor, ContentVisibilityEditor, TimedExaminationPreferenceEditor,
+        StaffLockEditor, UnitAccessEditor, DiscussionEditor, ContentVisibilityEditor, TimedExaminationPreferenceEditor,
```
Like the above comments, I think we can buck the cargo-cult trend here, remove this declaration (so we don't have to keep changing existing lines), and then just declare the variable below.

stvstnfrd replied:
```suggestion
        StaffLockEditor, UnitAccessEditor, ContentVisibilityEditor, TimedExaminationPreferenceEditor,
```

stvstnfrd commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -822,6 +822,74 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
         }
     });
 
+    DiscussionEditor = AbstractEditor.extend({
```
```suggestion
    var DiscussionEditor = AbstractEditor.extend({
```

stvstnfrd commented on cms/templates/js/discussion-editor.underscore:
```diff
@@ -0,0 +1,26 @@
+<form>
```
I'm guessing this is adapted from the existing editor templates?

Are there any key details (besides the content itself) that are worth calling out?

I'll punt on reviewing this file for now, as it looks like it will change a bit, following conversation on the related PR (to remove some of these UI indicators); let me know if/when I should review this in detail.

Agrendalath replied:
> I'm guessing this is adapted from the existing editor templates?
Are there any key details (besides the content itself) that are worth calling out?

Not really, you can just compare it to the [staff-lock-editor.underscore](https://github.com/edx/edx-platform/blob/1c90220462dfcc38096617c69796bb348289926d/cms/templates/js/staff-lock-editor.underscore), which I used as a template.

> I'll punt on reviewing this file for now, as it looks like it will change a bit, following conversation on the related PR (to remove some of these UI indicators); let me know if/when I should review this in detail.

Sure, will do.

stvstnfrd commented on cms/static/sass/elements/_modal-window.scss:
```diff
@@ -757,6 +758,7 @@
     }
 
     // UI: staff lock section
```
Is this comment no longer necessary/applicable?

Agrendalath replied:
I updated it. Should I remove it completely?

stvstnfrd commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -1069,6 +1137,11 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+                    {
+                        name: 'discussion',
+                        displayName: gettext('Discussion'),
+                        editors: [DiscussionEditor]
```
Per earlier conversation, would this block go away?
It looks like it's what's used to render the UI for non-verticals...

Agrendalath replied:
I'll get back to this comment after getting some more context on the other PR next week.

stvstnfrd commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -822,6 +822,74 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+            }
+        },
+
+        isIndeterminate: function() {
```
I suspect this (and other changes) are related to and/or may change following the discussion of simplifying the UI indicators, so unless you tell me otherwise, I'm going to punt on further reviewing this file for now.

Agrendalath commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -140,7 +141,9 @@ describe('CourseOutlinePage', function() {
             group_access: {},
-            user_partition_info: {}
+            user_partition_info: {
+                selectable_partitions: []
```
It wasn't an obvious error message in tests, but it turned out that `selectablePartitions` can be `undefined` [here](https://github.com/edx/edx-platform/blob/master/cms/templates/js/unit-access-editor.underscore#L6), so the unit tests related to unit's configuration were failing. I reverted this change and fixed it in a proper place now.

Agrendalath commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -11,10 +11,11 @@ import Course from 'js/models/course';
-        createMockSubsectionJSON, verifyTypePublishable, mockCourseJSON, mockEmptyCourseJSON, setSelfPaced,
-        mockSingleSectionCourseJSON, createMockVerticalJSON, createMockIndexJSON, mockCourseEntranceExamJSON,
-        selectOnboardingExam, createMockCourseJSONWithReviewRules,mockCourseJSONWithReviewRules,
+        selectVisibilitySettings, selectAdvancedSettings, selectDiscussionSettings, createMockCourseJSON,
```
Sure, that makes sense, thank you for pointing this out. I didn't see other reason than maintaining consistency here.

Agrendalath commented:
Thank you for the thorough review, @stvstnfrd.

> As noted, I've deferred review on some UI-sections, as it appears they'll change (significantly?) following discussions in the other PR (backend toggle).

I haven't been following these discussions too closely, so I will catch up on them next week. In the meantime I addressed your comments that were not strictly related to this.

> One piece I see missing is where is the site-wide toggle/flag for this feature?
Presumably, we don't want to show the editor options if we don't actually provide support for in-vertical discussions (or have a configured discussions provider).

>I think we'd need a flag like that at minimum to hide this until we have the supporting LMS/MFE UI built out, but maybe we also need to check if a discussions provider is available/enabled.

> It feels like we should at least be gating this with a feature flag for now; thoughts?

I'll defer this question to @xitij2000, as he is the BD-03 epic owner on our end.

xitij2000 commented:
For the site-wide toggle I was thinking that would be part of the new MFE, however, I guess it makes sense to have a system in the current UI as well. I guess having it in advanced configuration makes sense for now, while it's still not fully ready?

marcotuts commented:
Sorry for not catching this sooner - we should discuss the proposal here prior to continuing this work - I was not expecting any of this UI to be in Studio directly, but contained instead within a separate discussion settings area.

stvstnfrd commented:
@marcotuts @xitij2000 Assuming we still do not want to make these changes directly in Studio, should we close out this PR for now?

xitij2000 commented:
@stvstnfrd 
The current mockups show setting up "allow visibility configuration for each course unit" will this entirely be set up in the MFE?
The only reason for putting this in studio was that my assumption about the timeline meant that these changes would land before the MFE was ready, so the feature could be used pre-mfe deployment. Since the MFE has come up earlier, we can skip this.

stvstnfrd commented:
@xitij2000 Cool, yeah, especially since this still has the unresolved issue w/ the high query count in this implementation, I'm inclined to drop this for now and address it in the MFE (or manual config, eg Django Admin, for initial launch).

stvstnfrd commented:
(of course, hold on to the branch(es), in case it's useful later)

edx-webhook commented:
Hello @Agrendalath: We are unable to continue with review of your submission at this time. Please see the associated JIRA ticket for more explanation.
